### ISSUE TITLE
Encrypted swap devices do not work

### ISSUE THREAD

**Issue by DemiMarie** [systemd/systemd#12383]
**systemd version the issue has been seen with**
239

**Used distribution**
 Fedora 29

**Expected behaviour you didn't see**
System boots normally.

**Unexpected behaviour you saw**
Boot hangs.

**Steps to reproduce the problem**
Include a swap entry in `/etc/fstab` for an dm-crypt device that is generated by `/etc/crypttab` with `/dev/urandom` as entropy source.

**Comment by poettering** [comment]
https://freedesktop.org/wiki/Software/systemd/Debugging/#index1h1

Are you sure the device you list in /etc/crypttab exists?

**Comment by DemiMarie** [comment]
> https://freedesktop.org/wiki/Software/systemd/Debugging/#index1h1
> 
> Are you sure the device you list in /etc/crypttab exists?

It does.  However, systemd never gets any notification from udev with `SYSTEMD_READY=1`, so I had to write my own udev rules.  The culprit is that the device is encrypted with a freshly-generated key at each boot, so it is not recognized as a swap device by udev.

**Comment by poettering** [comment]
normally, the instant "mkswap" finishes and closes the block device, udev should get an inotify close event for that and then retrigger the device. Are you saying that doesn#t work?

**Comment by DemiMarie** [comment]
> normally, the instant "mkswap" finishes and closes the block device, udev should get an inotify close event for that and then retrigger the device. Are you saying that doesn#t work?

udev gets an event, but doesn’t recognize the device as a swap device.  It hits this udev rule:

```udevrules
SUBSYSTEM=="block", ENV{DM_UUID}=="CRYPT-*", ENV{ID_PART_TABLE_TYPE}=="", ENV{ID_FS_USAGE}=="", ENV{SYSTEMD_READY}="0"
```

**Comment by nfnty** [comment]
I have a similar issue, but boot hangs randomly waiting for the crypt device. About half the time it boots normally. This started happening after upgrading from 239 to 240. Never seen this issue before. Possibly related to #10179.

**Comment by bentolor** [comment]
Same issue here: boot hangs randomly waiting for the crypt device and I need to hit the "reboot" button of my VPS quite offen.

1 out of ~8 trials I got lucky and everything works fine. Really annoying.

**Comment by boz** [comment]
I'm running into this as well.

**Comment by eichin** [comment]
I may have a related case - ubuntu 16.04, `cryptswap1 UUID=... /dev/urandom swap,offset=1024,cipher=aes-xts-plain64`, hangs at `Started Cryptography Setup for cryptswap1.` with no timeout. Finally got it to happen with `systemd.debug-shell=1` turned on (in a VM with repeated clean reboots)  and was able to observe
```● systemd-cryptsetup@cryptswap1.service - Cryptography Setup for cryptswap1
   Loaded: loaded (/etc/crypttab; bad; vendor preset: enabled)
   Active: active (exited) since Thu 2019-07-04 21:16:34 EDT; 4h 48min ago
     Docs: man:crypttab(5)
           man:systemd-cryptsetup-generator(8)
           man:systemd-cryptsetup@.service(8)
  Process: 938 ExecStartPost=/sbin/mkswap /dev/mapper/cryptswap1 (code=exited, status=0/SUCCESS)
  Process: 760 ExecStart=/lib/systemd/systemd-cryptsetup attach cryptswap1 /dev/disk/by-uuid/fcf54d45-8755-479f-9213-15f9d357fdb2 /dev/urandom swap,offset=1024,cipher=aes-xts-plain64,loud (code=exited, status=0/SUCCESS)
 Main PID: 760 (code=exited, status=0/SUCCESS)
```
No relevant processes or cgroups were around.  Based on DemiMarie's comments above, I tried running `udevadm trigger` and suddenly it proceeded normally.  I'll see if I can get anything else from the udev side (I can generally reproduce the problem in less than an hour of reboots now.)

(Would love more advice on introspecting systemd (and I guess, udev) - this *was* a simple quick-install project for real hardware, and the hang-on-cryptsetup was intermittent enough that it was hard to tell if any given change had fixed it, plus the hang is early enough that `journald` didn't get anything to disk, the failed boots don't even show up in `--list-boots`.)

**Comment by eichin** [comment]
Indeed, `echo change >  /sys/dev/block/252:0/uevent` (where `/dev/mapper/cryptswap1 -> ../dm-0 -> 252, 0` is sufficient to unblock it, doesn't need any wider `trigger` than that.
![Screenshot_20190705_135237](https://user-images.githubusercontent.com/208492/60738789-87ab3580-9f2d-11e9-8056-1e031d33e879.png)

**Comment by DemiMarie** [comment]
My solution is to create `/etc/systemd/system/systemd-cryptsetup@.service.d/override.conf` with the contents:

```systemd
[Unit]
Before=dev-mapper-%i.swap
[Service]
ExecStartPost=/sbin/udevadm trigger /dev/mapper/%I
```

**Comment by 3point2** [comment]
Is this a duplicate of #10179?

**Comment by bentolor** [comment]
@3point2 For me  #10179 seems to match my situation: I experience this issue on a VPS. So the reporters assumpttion that it manly happens on slow/non-continuous IO/CPU is fully valid.

**Comment by renatoathaydes** [comment]
This happens for me on a modern Dell XPS (less than 2 years old) running Ubuntu 19, and there are several reports of people facing this problem in both virtual machines and (like me) physical machines.

See https://askubuntu.com/questions/341979/what-to-do-about-the-disk-drive-for-dev-mapper-cryptswap1-is-not-ready-yet-or#342002, https://askubuntu.com/questions/978043/swapon-stat-of-dev-mapper-cryptswap1-failed-no-such-file-or-directory and https://github.com/pop-os/pop/issues/316

**Comment by poettering** [comment]
Duplicate of #10179, which is fixed now. Closing
